### ISSUE TITLE
Prevent navigator/gecko.buildID leaks

### ISSUE THREAD

**Issue by nodiscc** [pyllyukko/user.js#117]
https://bugzilla.mozilla.org/show_bug.cgi?id=583181
https://www.browserleaks.com/javascript#comments
http://browserspy.dk/showprop.php

**Comment by CHEF-KOCH** [comment]
Random Agent Spoofer fixed this for me with script injection (which will not work anymore after FF 45 btw and needs to be fixed so don't try it on nightly's). 

![buuildid](https://cloud.githubusercontent.com/assets/727123/13058511/f8ea9138-d423-11e5-8294-ea53fd185af3.png)

**Comment by pyllyukko** [comment]
This is definitely a problem, yes. Firefox doesn't seem to respect ```gecko.buildID``` or any of the other settings :/

**Comment by nodiscc** [comment]
I can confirm that changing the `*.buildID` settings has no effect.

Random Agent Spoofer did not work for me, original Build ID is always visible to https://www.browserleaks.com/javascript

**Comment by pyllyukko** [comment]
Try now? :)

**Comment by CHEF-KOCH** [comment]
Info:
* https://wiki.mozilla.org/MozillaQualityAssurance:Build_Ids
* https://developer.mozilla.org/en/DOM/window.navigator.buildID
* http://kb.mozillazine.org/Locking_preferences

local-settings.js
```JavaScript
pref("general.config.obscure_value", 0);
pref("general.config.filename", "mozilla.cfg");
```

mozilla.cfg
``
lockPref("browser.startup.homepage_override.buildID", "yyyymmddnnnnnn");
lockPref("browser.sessionstore.upgradeBackup.latestBuildID", "yyyymmddnnnnnn");
lockPref("extensions.lastPlatformVersion", "nn.n");
lockPref("extensions.addonname.lastAppVersion", "nn.n");
lockPref("gecko.buildID", "yyyymmddnnnnnn");
lockPref("gecko.mstone", "nn.n");
lockPref("general.appname.override", "Netscape");
lockPref("general.appversion.override", "5.0 (Windows)");
lockPref("general.buildID.override", "yyyymmddnnnnnn");
lockPref("general.oscpu.override", "Windows NT 6.1");
lockPref("general.platform.override", "Win32");
lockPref("general.productSub.override", "20100101");
lockPref("general.useragent.appName", "Mozilla")
lockPref("general.useragent.override", "Mozilla/5.0 (Windows NT 6.1; rv:nn.n) Gecko/20100101 Firefox/nn.n");
lockPref("toolkit.telemetry.previousBuildID", "yyyymmddnnnnnn");
``

Can some#1 test if that works?

**Comment by nodiscc** [comment]
@pyllyukko your fix does not work for me

@CHEF-KOCH I can confirm that Random Agent spoofer does not affect the `Build ID` value on the test page for me (all script injection options enabled). I'm trying to locate where the lockPrefs file should reside, http://kb.mozillazine.org/Locking_preferences only describes locking for the Windows platform. I've tried `~/.mozilla/firefox/*/, /usr/share/iceweasel/{pref,profile}` without results.

Iceweasel 44.0.2-1 on Debian GNU/Linux stretch/sid

Please reopen?

**Comment by pyllyukko** [comment]
> Please reopen?

Will do.

**Comment by pyllyukko** [comment]
> @pyllyukko your fix does not work for me (original buildID still displayed)

That's odd.

> Iceweasel 44.0.2-1 on Debian GNU/Linux stretch/sid

Seems to work on my Debian 8 + Iceweasel 38.6.1. I wonder what's up?!

**Comment by CHEF-KOCH** [comment]
Well, that reminds me on our discussion we had in the past about the necessary, the page also needs Javascript enabled + cookies so again this is all nonsense if Javascript is disabled. So my question still stands without any one ever could tell me -- how is that related if someone not enabled javascript by default and only use a whitelist. Do I need all this on the pages 'I trust'. I guess no one can answer this.

> Random Agent spoofer does not affect the Build ID

Okay, strange it should work but I will open then a ticket to solve this.

**Comment by nodiscc** [comment]
I will retry with a freshly created profile.

**Comment by nodiscc** [comment]
Confirmed in newly created profile

```
$ apt-cache policy iceweasel
iceweasel:
  Installed: 45.0~b5-1
  Candidate: 45.0~b5-1
  Version table:
 *** 45.0~b5-1 700
          1 http://ftp.fr.debian.org/debian experimental/main amd64 Packages
        100 /var/lib/dpkg/status
     44.0.2-1 600
        500 http://ftp.fr.debian.org/debian stretch/main amd64 Packages
        200 http://ftp.fr.debian.org/debian unstable/main amd64 Packages
$ git clone https://github.com/pyllyukko/user.js
$ firefox --no-remote -P #create profile, exit
$ cp user.js/user.js ~/.mozilla/firefox/
$ cp user.js/user.js ~/.mozilla/firefox/*.mynewprofile/
$ firefox --no-remote -P #run with mynewprofile, go to http://browserspy.dk/showprop.php
#page shows buildID 20160214103951
```

May it be an iceweasel-specific bug?

**Comment by pyllyukko** [comment]
It works on my Iceweasel (see the User-Agent on the screenshot).

```
$ apt-cache policy iceweasel
iceweasel:
  Installed: 38.7.0esr-1~deb8u1
  Candidate: 38.7.0esr-1~deb8u1
  Version table:
 *** 38.7.0esr-1~deb8u1 0
        500 http://security.debian.org/ jessie/updates/main amd64 Packages
        100 /var/lib/dpkg/status
     38.5.0esr-1~deb8u2 0
        500 http://ftp.funet.fi/pub/linux/mirrors/debian/ jessie/main amd64 Packages
```

**Comment by nodiscc** [comment]
Fixed in Firefox 45.1.0esr-1 in Debian (pref `general.buildID.override` no actually overrides the buildId)
